File: leetcode/lc386.py
# An O(n) time and space solution (Note that the results is a list whose size
# is n, so this is the best prossible solution). However, it gets Time Limit
# Exceeded error on leetcode, which is totally crazy.

class Solution(object):
    def lexicalOrder(self, n):
        """
        :type n: int
        :rtype: List[int]
        """
        if n <= 0:
            return []

        rv = []

        def fill_rv(head_value=0):
            if head_value == 0:
                r = range(1, 10)
            else:
                r = range(10)
            promote_head_value = head_value * 10
            for i in r:
                new_value = promote_head_value + i
                if new_value > n:
                    return
                rv.append(new_value)
                fill_rv(new_value)

        fill_rv()
        return rv

if __name__ == "__main__":
    f = Solution().lexicalOrder
    print(f(13))
    print(f(23))
    f(4999999)

File: leetcode/lc386.cpp
#include <vector>

using namespace std;

class Solution {
    void lexicalOrder(int n, int val, vector<int>& rv) {
        int start = val==0?1:0;
        val *= 10;
        for (int i = start; i <= 9; ++i) {
            int temp = val + i;
            if (temp > n)
                return;
            rv.push_back(temp);
            lexicalOrder(n, temp, rv);
        }
    }
public:
    vector<int> lexicalOrder(int n) {
        vector<int> rv;
        lexicalOrder(n, 0, rv);
        return rv;
    }
};
